File: frontend/src/lib/modelCode.ts
// Код модели в конце названия: «… серебристый (Silver) (MDE54)» → MDE54.
// Единое правило с бэком (core/model_code.py и data-миграция v2w3x4y5z6a7): последняя
// скобочная группа из латиницы, цифр, дефиса и слэша длиной 3–12 символов, минимум одна
// цифра и хотя бы буква или дефис. «(2024)» (год) и «(Silver)» (цвет) кодом не считаются,
// «(492747-01)» (номер Dyson) — считается.
const CODE_TAIL = /\s*\(([A-Z0-9][A-Z0-9/-]{2,11})\)\s*$/

export function modelCodeFromName(name: string): string | null {
  const m = name.trimEnd().match(CODE_TAIL)
  if (!m) return null
  const code = m[1]
  return /\d/.test(code) && /[A-Z-]/.test(code) ? code : null
}

/** Название без кода в конце (если код там есть) */
export function stripModelCode(name: string): string {
  return modelCodeFromName(name) ? name.trimEnd().replace(CODE_TAIL, '').trimEnd() : name.trim()
}

/** Название с кодом в конце по правилу мастера групп: «база (КОД)»; без кода — просто база */
export function withModelCode(name: string, code: string | null): string {
  const base = stripModelCode(name)
  return code ? `${base} (${code})` : base
}

/** Ввод кода: латиница, цифры, дефис и слэш в верхнем регистре, до 20 символов */
export function normalizeModelCode(raw: string): string {
  return raw.toUpperCase().replace(/[^A-Z0-9/-]/g, '').slice(0, 20)
}

File: backend/src/app/core/model_code.py
"""Код модели в конце названия товара.

Владелец пишет код в конце названия: «… серебристый (Silver) (MDE54)». Правило одно
на всю систему — панель «Цены» на фронте (lib/modelCode.ts) и data-миграция
v2w3x4y5z6a7 повторяют его дословно: последняя скобочная группа из латиницы,
цифр, дефиса и слэша длиной 3–12 символов, минимум одна цифра и хотя бы одна
буква или дефис. Так «(MDE54)» и «(492747-01)» — код, а «(2024)» (год) и
«(Silver)» (цвет) — нет.
"""
from __future__ import annotations

import re

CODE_TAIL = re.compile(r"\s*\(([A-Z0-9][A-Z0-9/-]{2,11})\)\s*$")


def code_from_name(name: str | None) -> str | None:
    if not name:
        return None
    m = CODE_TAIL.search(name.rstrip())
    if not m:
        return None
    code = m.group(1)
    if not re.search(r"\d", code) or not re.search(r"[A-Z-]", code):
        return None
    return code


def strip_model_code(name: str | None) -> str:
    """Название без кода в конце (если код там есть)."""
    if not name:
        return ""
    if code_from_name(name) is None:
        return name.strip()
    return CODE_TAIL.sub("", name.rstrip()).rstrip()
